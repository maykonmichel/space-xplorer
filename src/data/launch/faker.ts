import faker from 'faker';
import {repeat} from 'ramda';

import {Launch} from '~/data/launch';

export const fakerFlickrImages = (): Launch['links']['flickr_images'] =>
  repeat(faker.image.image(), faker.datatype.number({min: 1, max: 5}));

export const fakerLaunch = (): Launch => ({
  id: faker.datatype.number().toString(),
  details: faker.lorem.paragraph(),
  mission_name: faker.lorem.words(),
  launch_date_formatted: faker.date.past().toLocaleDateString(),
  launch_date_local: faker.date.past().toLocaleDateString(),
  links: {
    article_link: faker.internet.url(),
    flickr_images: fakerFlickrImages(),
  },
  rocket: {
    rocket_name: faker.lorem.words(),
  },
});
