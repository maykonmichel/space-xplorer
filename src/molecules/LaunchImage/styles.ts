import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  image: {
    width: 200,
    height: 100,
  },
});
