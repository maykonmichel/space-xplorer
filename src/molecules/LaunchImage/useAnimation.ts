import {useEffect, useRef, useState} from 'react';
import {Animated} from 'react-native';

import colors from '~/theme/colors';

const duration = 500;
const inputRange = [0, 1];

export default (favorite: boolean) => {
  const value = favorite ? 'Remove from favorites' : 'Add to favorites';

  const [text, setText] = useState(value);

  useEffect(() => {
    const timout = setTimeout(() => setText(value), duration / 2);

    return () => clearTimeout(timout);
  }, [value]);

  const animatedValue = useRef(new Animated.Value(+favorite));

  useEffect(() => {
    Animated.timing(animatedValue.current, {
      duration,
      toValue: +favorite,
      useNativeDriver: false,
    }).start();
  }, [favorite]);

  const backgroundColor = animatedValue.current.interpolate({
    inputRange,
    outputRange: ['transparent', colors.secondary.main],
  });

  const color = animatedValue.current.interpolate({
    inputRange,
    outputRange: [colors.secondary.main, colors.neutral.light],
  });

  const opacity = animatedValue.current.interpolate({
    inputRange: [0, 0.5, 1],
    outputRange: [1, 0, 1],
  });

  const progress = animatedValue.current.interpolate({
    inputRange,
    outputRange: favorite ? [0.32, 0.45] : [0.2, 0],
  });

  return {
    backgroundColor,
    color,
    opacity,
    progress,
    text,
  };
};
