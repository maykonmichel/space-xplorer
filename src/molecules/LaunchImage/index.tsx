import React, {FC} from 'react';
import {TouchableOpacity} from 'react-native';
import FastImage from 'react-native-fast-image';

import styles from './styles';

export type Props = {
  onPress: (uri: string) => void;
  uri: string;
};

const LaunchImage: FC<Props> = ({onPress, uri}) => {
  const emit = () => onPress(uri);

  return (
    <TouchableOpacity onPress={emit} accessibilityLabel={'Toggle favorite'}>
      <FastImage
        source={{uri}}
        style={styles.image}
        accessibilityRole={'image'}
      />
    </TouchableOpacity>
  );
};

export default LaunchImage;
