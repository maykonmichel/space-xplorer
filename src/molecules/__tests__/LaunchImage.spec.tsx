import {fireEvent, render} from '@testing-library/react-native';
import faker from 'faker';
import React from 'react';

import LaunchImage, {Props} from '~/molecules/LaunchImage';

const getProps = (): Props => ({
  uri: faker.image.imageUrl(),
  onPress: jest.fn(),
});

describe('<LaunchImage />', () => {
  it('should render as expected', () => {
    const props = getProps();
    const {toJSON} = render(<LaunchImage {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should call onPress prop with uri arg on tap', () => {
    const props = getProps();
    const {uri, onPress} = props;
    const {getByA11yLabel} = render(<LaunchImage {...props} />);

    const touchable = getByA11yLabel('Toggle favorite');

    fireEvent.press(touchable);

    expect(onPress).toBeCalledWith(uri);
  });
});
