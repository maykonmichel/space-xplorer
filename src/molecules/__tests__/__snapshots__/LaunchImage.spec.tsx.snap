// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LaunchImage /> should render as expected 1`] = `
<View
  accessibilityLabel="Toggle favorite"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "height": 100,
          "width": 200,
        },
      ]
    }
  >
    <FastImageView
      accessibilityRole="image"
      resizeMode="cover"
      source={
        Object {
          "uri": "http://placeimg.com/640/480",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>
</View>
`;
